SU2_CFD/include/fluid/CCoolPropConductivity.hpp

```diff
@@ -0,0 +1,80 @@
+*
+* SU2 is free software; you can redistribute it and/or
+* modify it under the terms of the GNU Lesser General Public
+* License as published by the Free Software Foundation; either
+* version 2.1 of the License, or (at your option) any later version.
+*
+* SU2 is distributed in the hope that it will be useful,
+* but WITHOUT ANY WARRANTY; without even the implied warranty of
+* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+* Lesser General Public License for more details.
+*
+* You should have received a copy of the GNU Lesser General Public
+* License along with SU2. If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#pragma once
+
+
+#include "CConductivityModel.hpp"
+#include "CCoolProp.hpp"
+
+#ifdef USE_COOLPROP
+#include "CoolProp.h"
+#include "AbstractState.h"
+#endif
+/*!
+* \class CCoolPropConductivity
+* \brief Defines conductivity model from CoolProp.
+* \author P.Yan
+*/
+class CCoolPropConductivity final : public CConductivityModel {
+ private:
+#ifdef USE_COOLPROP
+  std::unique_ptr<CoolProp::AbstractState> fluid_entity;   /*!< \brief fluid entity */
+#endif
+  /*!< \brief list of fluids whose conductivity model is available in CoolProp */
+  vector<string> fluidNameList = {"Air", "Ammonia", "Argon", "Benzene", "CarbonDioxide", "Cyclopentane", "EthylBenzene",
+                              "Ethane", "Ethanol", "HeavyWater", "Helium", "Hydrogen", "IsoButane", "Isopentane",
+                              "Methane", "Methanol", "Nitrogen", "Oxygen", "Propylene", "ParaHydrogen", "R11", "R116" , "R12", "R123",
+                              "R1234yf", "R1234ze(E)", "R124", "R125", "R13", "R134a", "R14", "R141b", "R142b", "R143a", "R152A",
+                              "R218", "R22", "R227EA", "R23", "R236EA", "R236FA", "R245fa", "R32", "R404A", "R407C", "R410A",  "R507A",
+                              "RC318",  "SulfurHexafluoride", "Toluene", "Water", "m-Xylene", "n-Butane", "n-Decane", "n-Dodecane", "n-Heptane",
+                              "n-Hexane", "n-Nonane", "n-Octane", "n-Pentane", "n-Propane", "o-Xylene", "p-Xylene"};
+public:
+ /*!
+  * \brief Constructor of the class.
+  */
+ CCoolPropConductivity (string fluidname) {
+#ifdef USE_COOLPROP
+   fluid_entity = std::unique_ptr<CoolProp::AbstractState>(CoolProp::AbstractState::factory("HEOS",fluidname));
+   if (std::find(fluidNameList.begin(), fluidNameList.end(), fluidname) == fluidNameList.end()){
+       SU2_MPI::Error("Conductivity model not available for this fluid in CoolProp library.", CURRENT_FUNCTION);
+   }
+#else
+     SU2_MPI::Error("SU2 was not compiled with CoolProp (-Denable-coolprop=true). Note that CoolProp cannot be used with directdiff or autodiff", CURRENT_FUNCTION);
+#endif
+ }
+
+ /*!
+  * \brief Set thermal conductivity.
+  */
+ void SetConductivity(su2double t, su2double rho, su2double, su2double, su2double, su2double, su2double) override {
+#ifdef USE_COOLPROP
+   fluid_entity->update(CoolProp::DmassT_INPUTS, rho, t);
```

Comment:
Have you assessed the computational cost of CoolProp? When using fluid model + viscosity + conductivity, there are 3 identical fluid models being used.

Reply:
Yes. For the simulation of supersonic flow over a wedge with 38839 elements.

To run the simulation for 500 iteration:

case A: ideal gas model. Sutherland viscosity and conductivity model. It costs 20 second.
case B: CoolProp thermodynamics HEOS model, CoolProp viscosity and conductivity model. it costs around 10 min.
So the simulation is around 30 times slower. This effect is also reported in https://doi.org/10.1016/j.cpc.2019.107047

Reply:
I see. Do you know if that cost is associated more with the call to `update` or with the call to obtain the properties?
What I'm getting at is that it may be important to share the `fluid_entity` between the fluid, viscosity, and conductivity models.

Reply:
I agree with you. Actually,  I  tried this before. 
I created CCoolViscosity which is derived from both CViscoisty and CCoolProp, so that CCoolViscosity can access and use **fluid_entity** in CCoolProp.
But this results in an error in CFluidModel.cpp at:
"    case VISCOSITYMODEL::COOLPROP:
      return unique_ptr<CCoolPropViscosity>(new CCoolPropViscosity(config->GetFluid_Name()));"

So my opinion is that once this PR is merged, then I open a new PR to solve this issue and improve the simulation speed.

It is also a pity that in the mentioned paper, the authors do not share their source code, so that we can read for reference.

Reply:
Sounds good